diff --git a/src/CCA/Components/MPM/RigidMPM.cc b/src/CCA/Components/MPM/RigidMPM.cc
index a842076f6b..640dd48ad2 100644
--- a/src/CCA/Components/MPM/RigidMPM.cc
+++ b/src/CCA/Components/MPM/RigidMPM.cc
@@ -384,6 +384,7 @@ void RigidMPM::scheduleInterpolateToParticlesAndUpdate(SchedulerP& sched,
                       this, &RigidMPM::interpolateToParticlesAndUpdate);
 
   t->requiresVar(Task::OldDW, lb->delTLabel );
+  t->requiresVar(Task::OldDW, lb->simulationTimeLabel);
 
   Ghost::GhostType  gac = Ghost::AroundCells;
   t->requiresVar(Task::NewDW, lb->gTemperatureRateLabel,  gac,NGN);
@@ -469,6 +470,12 @@ void RigidMPM::interpolateToParticlesAndUpdate(const ProcessorGroup*,
     delt_vartype delT;
     old_dw->get(delT, lb->delTLabel, getLevel(patches) );
 
+    // Get the current simulation time
+    simTime_vartype simTimeVar;
+    old_dw->get(simTimeVar, lb->simulationTimeLabel);
+    double time = simTimeVar;
+
+
     for(unsigned int m = 0; m < numMPMMatls; m++){
       MPMMaterial* mpm_matl = (MPMMaterial*) m_materialManager->getMaterial( "MPM",  m );
       int dwi = mpm_matl->getDWIndex();
@@ -561,7 +568,15 @@ void RigidMPM::interpolateToParticlesAndUpdate(const ProcessorGroup*,
           tempRate += (gTemperatureRate[ni[k]] + dTdt[ni[k]])   * S[k];
           acc      += gacceleration[ni[k]]   * S[k];
         }
-        pTempNew[idx]        = pTemperature[idx] + tempRate  * delT;
+        if(m==0){
+          pTempNew[idx]    = pTemperature[idx] + tempRate*delT;
+        } else {
+          if(time<=60.){
+            pTempNew[idx]    = 1644.;
+          } else {
+            pTempNew[idx]    = 300.;
+          }
+        }
         pvelocitynew[idx]    = pvelocity[idx]+acc*delT;
         // If there is no adiabatic heating, add the plastic temperature
         // to the particle temperature
