*** tk8.3.4/unix/configure	Fri Oct 19 18:24:40 2001
--- tk8.3.4/unix/configure-patch	Thu Jan 26 14:06:06 2006
*************** echo "configure:1679: checking system ve
*** 1688,1694 ****
  	    # results, and the version is kept in special file).
  	
  	    if test -r /etc/.relid -a "X`uname -n`" = "X`uname -s`" ; then
! 		system=MP-RAS-`awk '{print }' /etc/.relid'`
  	    fi
  	    if test "`uname -s`" = "AIX" ; then
  		system=AIX-`uname -v`.`uname -r`
--- 1688,1694 ----
  	    # results, and the version is kept in special file).
  	
  	    if test -r /etc/.relid -a "X`uname -n`" = "X`uname -s`" ; then
! 		system=MP-RAS-`awk '{print }' /etc/.relid`
  	    fi
  	    if test "`uname -s`" = "AIX" ; then
  		system=AIX-`uname -v`.`uname -r`
*** tk8.3.4/unix/tcl.m4	Fri Oct 19 18:24:40 2001
--- tk8.3.4/unix/tcl.m4-patch	Thu Jan 26 14:06:19 2006
*************** AC_DEFUN(SC_CONFIG_CFLAGS, [
*** 584,590 ****
  	    # results, and the version is kept in special file).
  	
  	    if test -r /etc/.relid -a "X`uname -n`" = "X`uname -s`" ; then
! 		system=MP-RAS-`awk '{print $3}' /etc/.relid'`
  	    fi
  	    if test "`uname -s`" = "AIX" ; then
  		system=AIX-`uname -v`.`uname -r`
--- 584,590 ----
  	    # results, and the version is kept in special file).
  	
  	    if test -r /etc/.relid -a "X`uname -n`" = "X`uname -s`" ; then
! 		system=MP-RAS-`awk '{print $3}' /etc/.relid`
  	    fi
  	    if test "`uname -s`" = "AIX" ; then
  		system=AIX-`uname -v`.`uname -r`
*************** AC_DEFUN(SC_BLOCKING_STYLE, [
*** 1708,1714 ****
  	    # results, and the version is kept in special file).
  	
  	    if test -r /etc/.relid -a "X`uname -n`" = "X`uname -s`" ; then
! 		system=MP-RAS-`awk '{print $3}' /etc/.relid'`
  	    fi
  	    if test "`uname -s`" = "AIX" ; then
  		system=AIX-`uname -v`.`uname -r`
--- 1708,1714 ----
  	    # results, and the version is kept in special file).
  	
  	    if test -r /etc/.relid -a "X`uname -n`" = "X`uname -s`" ; then
! 		system=MP-RAS-`awk '{print $3}' /etc/.relid`
  	    fi
  	    if test "`uname -s`" = "AIX" ; then
  		system=AIX-`uname -v`.`uname -r`
